crates/full-node/sov-sequencer/src/preferred/timestamp.rs

```diff
@@ -0,0 +1,153 @@
+use crate::common::Sequencer;
+use crate::preferred::PreferredSequencer;
+use crate::preferred::Runtime;
+use crate::preferred::UniquenessData;
+use borsh::BorshSerialize;
+use derive_more::FromStr;
+use full_node_configs::sequencer::TimingOracleConfig;
+use sov_modules_api::capabilities::TransactionAuthenticator;
+use sov_modules_api::macros::config_value;
+use sov_modules_api::transaction::PriorityFeeBips;
+use sov_modules_api::transaction::TxDetails;
+use sov_modules_api::transaction::UnsignedTransaction;
+use sov_modules_api::Amount;
+use sov_modules_api::CryptoSpec;
+use sov_modules_api::HexString;
+use sov_modules_api::PrivateKey;
+use sov_modules_api::PublicKey;
+use sov_modules_api::RawTx;
+use sov_modules_api::Spec;
+use sov_rollup_interface::node::da::DaService;
+use std::sync::Arc;
+use tokio::sync::watch;
+use tokio::time::Duration;
+
+pub(crate) struct TimingOracleConfigWithPrivateKey<S: Spec> {
+    time_oracle_config: TimingOracleConfig,
+    priv_key: <S::CryptoSpec as CryptoSpec>::PrivateKey,
+}
+
+impl<S: Spec> TimingOracleConfigWithPrivateKey<S> {
+    pub(crate) fn new(
+        maybe_time_oracle_config: Option<TimingOracleConfig>,
+    ) -> Option<anyhow::Result<Self>> {
+        let time_oracle_config = maybe_time_oracle_config?;
+        let priv_key = || -> anyhow::Result<_> {
+            match &time_oracle_config.private_key_hex {
+                Some(key_str) => {
+                    let bytes = HexString::from_str(key_str).map_err(|_| anyhow::anyhow!("Invalid oracle private key hex - could not parse as hex. Check your preferred sequencer config file."))?.0;
+                    let key = <S::CryptoSpec as CryptoSpec>::PrivateKey::try_from(bytes).map_err(|_| anyhow::anyhow!("Invalid oracle private key hex - invalid private key bytes. Check your preferred sequencer config file."))?;
+                    Ok(key)
+                }
+                None => {
+                    let key = <S::CryptoSpec as CryptoSpec>::PrivateKey::generate();
```

Comment:
This will only work if the rollup has a `paymaster` configured; otherwise, the new address won’t have any funds. I added a warning about this. If we remove the ephemeral key functionality the only downside is that we’d need to put the private key in the config right?

Reply:
Yeah, correct